python/dgl/data/csv_dataset.py

```diff
@@ -0,0 +1,427 @@
+                assign_data(type, data['data'], graph.edges)
+            graph_dict[graph_id] = graph
+        return graph_dict
+
+
+class DefaultDataParser:
+    """ Default data parser for DGLCSVDataset. It
+        1. ignores any columns which does not have a header.
+        2. tries to convert to list of numeric values(generated by
+            np.array().tolist()) if cell data is a str separated by ','.
+        3. read data and infer data type directly, otherwise.
+    """
+
+    def __call__(self, df: pd.DataFrame):
+        data = {}
+        for header in df:
+            if 'Unnamed' in header:
+                dgl_warning("Unamed column is found. Ignored...")
+                continue
+            dt = df[header].to_numpy().squeeze()
+            if len(dt) > 0 and isinstance(dt[0], str):
+                #probably consists of list of numeric values
+                dt = np.array([ast.literal_eval(row) for row in dt])
+            data[header] = dt
+        return data
+
+
+class DGLCSVDataset(DGLDataset):
+    """ This class aims to parse data from CSV files, construct DGLGraph
+        and behaves as a DGLDataset.
+
+    Parameters
+    ----------
+    data_path : str
+        Directory which contains 'meta.yaml' and CSV files
+    force_reload : bool
+        Whether to reload the dataset. Default: False
+    verbose: bool
+        Whether to print out progress information. Default: True.
+    node_data_parser : dict
+        A dictionary used for node data parsing when loading from CSV files.
+        The key is node type which specifies the header in CSV file and the
+        value is a callable object which is used to parse corresponding
+        column data. Default: None. If None, a default data parser is applied
+        which load data directly and tries to convert list into array.
+    edge_data_parser : dict
+        A dictionary used for edge data parsing when loading from CSV files.
+        The key is edge type which specifies the header in CSV file and the
+        value is a callable object which is used to parse corresponding
+        column data. Default: None. If None, a default data parser is applied
+        which load data directly and tries to convert list into array.
+    graph_data_parser : callable
+        A callable object which is used to parse corresponding column graph
+        data. Default: None. If None, a default data parser is applied
+        which load data directly and tries to convert list into array.
+
+    Attributes
+    ----------
+    graphs : :class:`dgl.DGLGraph`
+        Graphs of the dataset
+    data : dict
+        any available graph-level data such as graph-level feature, labels.
+
+    Examples
+    --------
+    Single heterograph dataset
+    >>> csv_dataset = data.DGLCSVDataset(test_dir)
+    >>> g = csv_dataset[0]
+
+    Multiple graphs dataset
+    >>> csv_dataset = data.DGLCSVDataset(test_dir)
+    >>> g, label = csv_dataset[0]
```

Comment:
These examples are not really useful and I understand it is difficult to show the example without showing the data file content. Maybe it's better just have a link to some docpage?

Reply:
as discussed offline, let's remove examples from here and add a new/detailed web page to elaborate how to use `DGLCSVDataset`.